web/api/rest-api-design/Batching-Filtering-Pagination.md

```diff
@@ -0,0 +1,73 @@
+## Batching, Filtering, and Pagination
+
+### Filtering:
+
+- Topics covered: Elasticsearch and URL Parameters
+
+#### Elasticsearch:
+
+- Elasticsearch is a distributed RESTful search engine that is recommended for most filtering needs.
+- It works by sorting results by a relevance score, which is a way to quantify how well each result matches a query (higher score = more relevance, e.g. the more relevant the document being searched)
+- The score is also determined by whether the query is run via a query or filter context.
+- In a **query context** the following query clause is passed to a query parameter to calculate a relevance score: How well does this document match this query clause?
+- When a **filter context** is used, which happens mostly when filtering structured data, a query clause is passed to a filter parameter (e.g. `filter`). Some examples include:
+
+  - Does this timestamp fall into the range 2015 to 2016?
+  - Is the status field set to "published"?
+
+Examples of a query and filter context:
+
+```
+GET /_search
+{
+  "query": {
+    "bool": {
+      "must": [
+        { "match": { "title":   "Search" }},
+        { "match": { "content": "Elasticsearch" }}
+      ],
+      "filter": [
+        { "term":  { "status": "published" }},
+        { "range": { "publish_date": { "gte": "2015-01-01" }}}
+      ]
+    }
+  }
+}
+
+```
+
+In this example we see the query parameter `”query”:{...}` indicates a context. The `bool` and two `match clauses` within the query context are used to score the document relevance. Finally, the filter parameters `”filter”: [...]` have two clauses, `term` and `range`, which are part of the filter context and allow it to filter out the documents that don’t match.
+
+#### URL parameters:
+
+- You can filter using resource names. For example, if you have an API that tracks different dogs breeds, their names, and their birth years.
+
+`GET https://api.thedogapi.com/v1/breeds`
+
+- If adding a query to your url parameters, use question marks before providing a query parameters, e.g. limits:
+  `GET https://api.thedogapi.com/v1/breeds?limit=2`
+- Square brackets on key names within a resource provide more flexibility with filtering and enable you to indicate a range. For example, one way to indicate a range of values:
+
+  If you want to search for dogs born between 2019 and 2021:
+  `GET https://api.thedogapi.com/v1/dogs/birth[year]=2019&birth[year]=2021`
+
+  If you want to search for dogs whose birthdays are in the Spring you can search using a range of months:
+  `GET https://api.thedogapi.com/v1/dogs/birth[month]=March&birth[month]=May`
+
+- The above examples using brackets indicate different resources that can be passed in as URL parameters to indicate range.
+
+- Search Query Params allow for more flexibility in adding filters to a URL. You can add sorts and other search parameters to make filtering suit your needs more efficiently
+
+#### Downside to URL Parameters in Filtering:
```

Comment:
I would not consider most of these downsides to URL parameters. URL parameters themselves do not perform any filtering. The filtering happens on the backend when it interprets the query parameters. You can design the query parameters however you want, in a way that allows the end user to perform rich queries. 

GitHub's API is a great example. Here is an example query that searches for repos owned by the `github` org OR by `trussworks`, written in Ruby, and with followers in the range of 10 to 1000:

```
https://github.com/search?q=user%3Agithub+user%3Atrussworks+language%3ARuby+followers%3A10..1000
```

Note how GitHub designed their advanced search to use the `q` parameter and they use a colon (`%3A`) to assign values to the various search fields, and a `+` to add different fields to the search. Ranges are identified with two dots.

As this example shows, you can design your query parameters in any way you want, and then it's up to the backend to parse the query and perform the right search in the backend, using your search tool of choice.